### DISCUSSION TITLE
Using C# to retrieve and then apply retention labels

### DISCUSSION THREAD

**Discussion by przeciag** [pnp/pnpframework#304] [Q&A]
Hi everyone, I'm wondering if someone could point me in the right direction. I've got to copy retention labels from one library directory into another. Could someone help and point me in the right direction with the following:
1. Access files in list (I can do that in powershell, but struggle to do it in C#)
2. Retrieve and apply retention labels

Thanks in advance

**Comment by jansenbe** [comment]
@przeciag : PnP Core SDK (our new .NET library, used by PnP Framework) does have support for reading and setting compliance tags on lists. You can get a PnPContext object from a given CSOM ClientContext like this:

```csharp
var pnpContext = PnPCoreSdk.Instance.GetPnPContext(web.Context as ClientContext);
```

Once you have a PnPContext instance you can get your list(s) as explained in https://pnp.github.io/pnpcore/using-the-sdk/lists-intro.html#getting-lists and then use one of the below methods:

![image](https://user-images.githubusercontent.com/7451219/116870823-84f5f600-ac13-11eb-924c-9c2d0a336e1f.png)

**Reply by jansenbe** [reply]
See also https://pnp.github.io/pnpcore/using-the-sdk/basics-pnpframework.html for PnP Core SDK interop

**Reply by przeciag** [reply]
@jansenbe this is all very helpful thank you! I've managed to retrieve lists from sharepoint, and can iterate through all of them with no problem. Methods you mentioned before seem to only apply to lists, is it possible to apply retention tags to specific files in a list?

**Reply by jansenbe** [reply]
@przeciag : at this moment we don't have support for doing that in PnP Core SDK, this is something we however should implement. I've created https://github.com/pnp/pnpcore/issues/417 to track this request. Thanks for making us aware. If you can use CSOM you can still use that (the examples in the issue I've created are CSOM). If you don't want a CSOM dependency then please wait for the change or ideally update PnP Core SDK yourselves and create a PR so that we can add your implementation :-)
